bam_reheader.c

```diff
@@ -453,30 +467,104 @@ int main_reheader(int argc, char *argv[])
     int inplace = 0, r, add_PG = 1, c;
     bam_hdr_t *h;
     samFile *in;
-    char *arg_list = stringify_argv(argc+1, argv-1);
+    char *arg_list = stringify_argv(argc+1, argv-1), *external = NULL;
 
     static const struct option lopts[] = {
         {"help",     no_argument, NULL, 'h'},
         {"in-place", no_argument, NULL, 'i'},
         {"no-PG",    no_argument, NULL, 'P'},
+        {"external", required_argument, NULL, 'e'},
         {NULL, 0, NULL, 0}
     };
 
-    while ((c = getopt_long(argc, argv, "hiP", lopts, NULL)) >= 0) {
+    while ((c = getopt_long(argc, argv, "hiPe:", lopts, NULL)) >= 0) {
         switch (c) {
         case 'P': add_PG = 0; break;
         case 'i': inplace = 1; break;
+        case 'e': external = strdup(optarg); break;
         case 'h': usage(stdout, 0); break;
         default:
             fprintf(stderr, "Invalid option '%c'\n", c);
             usage(stderr, 1);
         }
     }
 
-    if (argc - optind != 2)
+    if ((argc - optind != 2 || external) && (argc - optind != 1 || !external))
         usage(stderr, 1);
 
-    { // read the header
+    if (external) {
+        in = sam_open(argv[optind], inplace?"r+":"r");
```

Comment:
If you open the file with `hopen()` followed by `hts_hopen()`, you will already have the `hFILE*` and won't have a desire for an `hts_hfile()` function.